OAK-10803 - Fix memory consumption of uncompress properties

ionutzpi opened the pull request:
In [OAK-10803](https://issues.apache.org/jira/browse/OAK-10803) we introduce the possibility of compressing property values. During its development we encountered performance issues, namely with repeated decompression calls, even just during getNodeAtRevision. Before we can confidentially enable and use compression, we need to revisit performance first. Likely we'll have to make further tweaks before we can use it.

Commit:
OAK-10973 - Performance tune property compression/decompression

reschke reviewed (commented):
This saves space, but instead will make every read access slower, right?

Commit:
OAK-10973 - Performance tune property compression/decompression

ionutzpi commented:
Done. Refactored it to make read access faster.

stefan-egli commented:
I don't see the value in this. This will restrict usability of compression to the time between creation and first access of the value. Thereafter it uses more space.

Commit:
OAK-10973 - Performance tune property compression/decompression revert

ionutzpi commented:
Revert it to first commit. Yes, the read access is slower.

stefan-egli commented:
`decompressedValue` is never set?

Commit:
OAK-10973 - Performance tune property compression/decompression set decompressedValue

ionutzpi commented:
Done.

Commit:
OAK-10973 - Performance tune property compression/decompression set decompressedValue

Commit:
OAK-10973 - Performance tune property compression/decompression separate DocumentPropertyState from CompressedDocumentPropertyState

stefan-egli commented:
IIUC this PR has now addressed the [suggestion](https://issues.apache.org/jira/browse/OAK-10803?focusedCommentId=17869342&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17869342) on OAK-10803. That I think we should follow-up with indeed.

What about doing it in 2 steps though:
* first address those suggestions (bring memory consumption back to pre-compression)
* then look into performance improvements

Currently this PR seems to mix both concerns, which makes review discussion a bit more complex.

Having said that, reg the performance improvements: I still think we need to address the performance aspect differently. As it stands now, the first call to `decompress()` will expand the value again to its original state (I would then have perhaps set `value` instead of introducing/duplicating `decompressedValue`) - which means it will use up again the original amount of memory (at which point the gain done by compression is over). The issue is that `decompress()` will be call pretty much immediately after a property was created - namely when it needs to be put into the cache and when its memory consumption is estimated. Hence there would be zero gain of compression if decompression was done as it stands now. (we could for example verify how the memory consumption calculation could be fixed - after all it's probably broken now with compression anyway - and perhaps that could make the compression state live longer)

Commit:
OAK-10973
 separate DocumentPropertyState from CompressedDocumentPropertyState delete decompressedValue

Commit:
OAK-10973
 separate DocumentPropertyState from CompressedDocumentPropertyState refactor tests

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateFactory.java:
```diff
@@ -0,0 +1,15 @@
+package org.apache.jackrabbit.oak.plugins.document;
```
license header is missing

ionutzpi replied:
Done.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateFactory.java:
```diff
@@ -0,0 +1,15 @@
+
+public class DocumentPropertyStateFactory {
+
+    public static PropertyState createPropertyState(DocumentNodeStore store, String name, String value, Compression compression) {
```
what about adding a variant that doesn't ask for the `Compression`? in that case it would perhaps hard code to GZIP, or later that could be configured..

ionutzpi replied:
Done

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,354 @@
+        this.compression = compression;
+        int size = value.length();
+        byte[] compressedValue = null;
+        if (size > COMPRESSION_THRESHOLD) {
```
I would suggest to move that out of the constructor and check the size in the factory. Would also simplify the code of this class

ionutzpi replied:
Done

stefan-egli reviewed (commented):
Is the plan to re-purpose this PR to now address OAK-10803? Then we might want to change the subject of the PR (that usually also becomes the commit message later).

Commit:
OAK-10973
 separate DocumentPropertyState from CompressedDocumentPropertyState refactor factory method

ionutzpi commented:
I separated construction of DocumentPropertyState from CompressedDocumentPropertyState through a factory method for step 1: first address those suggestions (bring memory consumption back to pre-compression)
We can change the subject

stefan-egli reviewed (commented):
What's missing is the wiring. With this change, the compression code is actually never triggered since it doesn't go through the factory yet.

Commit:
OAK-10973
 call DocumentPropertyState with factory method

ionutzpi commented:
Done. Call DocumentPropertyState through factory method.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateFactory.java:
```diff
@@ -0,0 +1,46 @@
+
+    public static PropertyState createPropertyState(DocumentNodeStore store, String name, String value, Compression compression) {
+
+        if (compression != null && !compression.equals(Compression.NONE) && value.length() > CompressedDocumentPropertyState.getCompressionThreshold()) {
```
the length check here would result to true for `-1` - shouldn't there be a test for that, did that not fail for some reason?

ionutzpi replied:
The tests are checking default threshold exceeded or not, it didnt failed because of passing Compression.NONE as argument

stefan-egli replied:
There should be a test that should fail now though. The condition is currently wrong. 

Probably we should also use GZIP by default, otherwise the feature cannot be enabled at all?

ionutzpi replied:
Done. We can enable the feature by calling differently factory method.

stefan-egli replied:
> We can enable the feature by calling differently factory method.

But that would require a code change. What I was referring to is enabling the feature by configuration - eg via the system property. We should have that - otherwise the code is not useable and disable in a hard-coded way. I do think we should use GZIP in the default, not NONE.

ionutzpi replied:
Done.

stefan-egli replied:
The test still doesn't fail consistently. I'm referring to `DocumentPropertyStateFactoryTest.createPropertyStateWithDefaultCompression`. That is one test at least that should fail with the current code. The reason it doesn't fail is that it depends on test execution order. Some other test method in that class is executed first (eg `createPropertyStateWithCompressionThresholdNotExceeded`) and that sets the threshold, so that `createPropertyStateWithDefaultCompression` then runs on false assumptions.

Things that require fixing:
* createPropertyStateWithDefaultCompression should check that it actually uses the default - i.e. it should have an assertEquals before even executing the actual property creation (and that would currently fail, depending on test execution order)
* in addition to such asserts in each method, there must also be a `@After` and a `@Before` that resets the threshold
* plus the actual code fix

ionutzpi replied:
Done

Commit:
OAK-10973 -- added tests for DocumentPropertyStateFactoryTest

Commit:
OAK-10973 -- added tests for DocumentPropertyStateFactoryTest

Commit:
OAK-10973 -- added another test method

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,350 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * PropertyState implementation with lazy parsing of the JSOP encoded value.
```
Would suggest to add some short mention of the fact that this class uses compression.

ionutzpi replied:
Done

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,350 @@
+        return createProperty(name, values, Type.fromTag(type, true));
+    }
+}
+
+
+
+
+
+
+
```
detail but we probably should remove those lines

ionutzpi replied:
Done

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java:
```diff
@@ -47,64 +38,24 @@
-        this(store, name, value, Compression.GZIP);
-    }
-
-    DocumentPropertyState(DocumentNodeStore store, String name, String value, Compression compression) {
```
Presumably this class should be reverted to the state before compression? In doing that it looks like it missed a commit that came in between : b04de7ea9dce0859de09f90964fa650a7bf5a978

ionutzpi replied:
Manually added that commit

stefan-egli reviewed (commented):
Haven't yet fully checked but it looks like some tests got lost in translation? eg multiValuedAboveThresholdSize. Also, compressValueThrowsException : now expects an exception while previously it didn't?

reschke commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java:
```diff
@@ -0,0 +1,168 @@
+    @Rule
+    public DocumentMKBuilderProvider builderProvider = new DocumentMKBuilderProvider();
+
+    private Set<String> reads = newHashSet();
```
please now new use of Guava.

ionutzpi replied:
Done

Commit:
OAK-10973 -- set default GZIP for compression

Commit:
OAK-10973 -- refactor

Commit:
OAK-10973 -- manually added b04de7ea9dce0859de09f90964fa650a7bf5a978 revision

Commit:
OAK-10973 -- replace guava collection with jdk one

ionutzpi commented:
The constructor in CompressedDocumentPropertyState has changed from the previous version of DocumentPropertyState

Commit:
OAK-10973 -- refactor tests for default Compression GZIP

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,343 @@
+     */
+    static PropertyState readArrayProperty(String name, DocumentNodeStore store, JsopReader reader) {
+        int type = PropertyType.STRING;
+        List<Object> values = Lists.newArrayList();
```
I think this can be refactored too, and then entirely avoid the import of the guava dependency. I think that's what @reschke wanted to point out.

ionutzpi replied:
Done. It wasn't very clear from the comment.

Commit:
OAK-10973 -- refactor guava collection

stefan-egli commented:
> The constructor in CompressedDocumentPropertyState has changed from the previous version of DocumentPropertyState

What question or comment does this refer to?

ionutzpi commented:
> Also, compressValueThrowsException : now expects an exception while previously it didn't?

This question

stefan-egli commented:
(Adding a general comment here in the main discussion as it otherwise might be getting a bit complicated by now)

What I was trying to say: the fact that the `DocumentPropertyStateTest` got refactored into 3 different classes lost a few aspects and thus results in a regression.
* as mentioned I believe some test methods got lost (I will double-check and can add more details)
* in addition though, some of the `@Before` and `@After` logic got lost - which actually is also a regression. For example, the broken surrogate tests no longer work as they were intended - they now inherit the threshold from whatever test/test-method was executed before, which is breaking change from the previous behavior (I think previously they tested with `-1`)
* (and the `-1` bug mentioned several times is thus still undetected by tests and is still existing in the code)

Plus what is also a regression vs the previous state is what I mentioned previously:
> Also, compressValueThrowsException : now expects an exception while previously it didn't?

Previously the constructor was swallowing those exceptions, which I think is a good thing. Now that has been changed. Why?

Commit:
OAK-10973 -- refactor guava collection and move test broken surrogate to CompressedDocumentPropertyStateTest

ionutzpi commented:
Moved broken surrogate tests on CompressedDocumentPropertyStateTest(before and after are in place here).
In CompressDocumentPropertyState constructor the logic is only for compression, if it fails then throw new IllegalArgumentException.

stefan-egli commented:
> In CompressDocumentPropertyState constructor the logic is only for compression, if it fails then throw new IllegalArgumentException.

That's a change to the previous behavior though. We previously discussed that it shouldn't throw an exception for compressing a property but just fall back to uncompressed. What is the reason for this change?

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java:
```diff
@@ -113,287 +80,4 @@ public void multiValuedBinarySize() throws Exception {
-        verify(mockCompression, times(1)).getInputStream(any(InputStream.class));
-    }
-
-    @Test
-    public void defaultValueSetToMinusOne() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException, NoSuchFieldException {
-        DocumentNodeStore store = mock(DocumentNodeStore.class);
-
-        DocumentPropertyState.setCompressionThreshold(-1);
-        DocumentPropertyState state = new DocumentPropertyState(store, "propertyNameNew", "\"" + STRING_HUGEVALUE + "\"", Compression.GZIP);
-
-        byte[] result = state.getCompressedValue();
-
-        assertNull(result);
-        assertEquals(state.getValue(Type.STRING), STRING_HUGEVALUE);
-    }
```
This test seems removed. Is there a replacement for it?

ionutzpi replied:
Done. I added a replacement

ionutzpi replied:
In CompressedDocumentPropertyState constructor if it throws an exception, it will be caught in factory method and call DocumentPropertyState constructor.

Commit:
OAK-10973 -- added -1 condition into factory method

Commit:
OAK-10973 -- added test for -1

stefan-egli commented:
> In CompressedDocumentPropertyState constructor if it throws an exception, it will be caught in factory method and call DocumentPropertyState constructor.

(Moving this back to main thread)

Could the test (`compressValueThrowsException`) then be adjusted to use the factory and thus avoid `(expected = IllegalArgumentException.class)` ?

stefan-egli commented:
On the test methods that seem to have gotten lost, here's a list of them - those existed prior to this PR and now seem gone (unless renamed) :
```
uncompressValueThrowsException
multiValuedAboveThresholdSize
stringAboveThresholdSizeNoCompression
testEqualsWithoutCompression
testInterestingStringsWithoutCompression
testOneCompressOtherUncompressInEquals
uncompressValueThrowsException
```

Also noticed that `stringBelowThresholdSize` actually tests _below_ threshold - so either the test name or the test code is wrong.

Also, as a general comment : I think it would be useful to go via the factory method whenever possible. That would extend the test coverage of the factory method - which is a very key element. (I think this would be the lost prio comment though)

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyState.java:
```diff
@@ -19,24 +19,16 @@
 
 import javax.jcr.PropertyType;
 
 import org.apache.jackrabbit.guava.common.collect.Lists;
```
we might want to remove this guava usage as well, while we're effectively changing this class entirely.

ionutzpi replied:
Done.

ionutzpi commented:
All the tests mention above were refactored to use new context. The test suite created covered all scenarios previous done.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,343 @@
+/**
+ * PropertyState compression implementation with lazy parsing of the JSOP encoded value.
+ */
+public final class CompressedDocumentPropertyState implements PropertyState {
```
I think this class could deserve some refactoring. The way it is now it basically duplicates a lot of code from DocumentPropertyState, which is not ideal.

Perhaps the reason for not directly extending DocumentPropertyState is that would inherit `value` and thus use more memory?

Then again I'm not liking the current code duplication a lot neither.

I think at least we should try to reuse what's easily possible - namely the static methods. Those we don't need in this class as well and we can use the DocumentPropertyState one.

wdyt?

ionutzpi replied:
Done. Remove duplicate static methods. 
The 2 classes have different logic for calculating value, and also use more memory if subclassing, it should be separated.

Commit:
OAK-10973 -- remove guava from DocumentPropertyState

stefan-egli commented:
> All the tests mention above were refactored to use new context. The test suite created covered all scenarios previous done.

where is uncompressValueThrowsException covered? where testInterestingStringsWithoutCompression ?

edit: plus where is testOneCompressOtherUncompressInEquals covered?

Commit:
OAK-10973 -- remove duplicate static methods in CompressedDocumentPropertyState

Commit:
OAK-10973 -- added required test methods CompressedDocumentPropertyStateTest

ionutzpi commented:
Added required test methods in CompressedDocumentPropertyStateTest

Commit:
OAK-10973 -- refactor if statement

Commit:
OAK-10973 -- refactor if statement

Commit:
OAK-10973 -- refactor test methods

Commit:
OAK-10973 -- refactor test methods

Commit:
OAK-10973 -- refactor test methods

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java:
```diff
@@ -0,0 +1,243 @@
+ */
+package org.apache.jackrabbit.oak.plugins.document;
+
+import static java.util.Collections.emptyList;
+import static org.apache.jackrabbit.oak.plugins.memory.PropertyStates.createProperty;
```
looking good now - except I'd suggest to remove the unused imports here (in this class), then we should be good to go

ionutzpi replied:
Done. Thanks

Commit:
OAK-10973 -- remove unused imports

stefan-egli reviewed (approved):
+1, looks good now. Will let the test run finish (but I think 2 tests also fail in trunk, so those should anyway be ignored), and then (squash) merge depending on results.

stefan-egli commented:
ok, test failed, but the failures looks like unrelated flaky ones. going to merge anyway.
